pkg/observable/notifiable/observable.go

```diff
@@ -0,0 +1,93 @@
+package notifiable
+
+import (
+	"context"
+	"sync"
+
+	"pocket/pkg/observable"
+)
+
+// Observable implements the observable.Observable interface and can be notified
+// via its corresponding notifier channel.
+type Observable[V any] struct {
+	mu          sync.RWMutex
+	ch          <-chan V // private channel that is used to emit values to subscribers
+	subscribers []chan V // subscribers is a list of channels that are subscribed to the observable
+	closed      bool
+}
+
+// NewObservable creates a new observable is notified when the notifier channel
+// receives a value.
+func NewObservable[V any](notifier chan V) (observable.Observable[V], chan V) {
+	// If the caller does not provide a notifier, create a new one and return it
+	if notifier == nil {
+		notifier = make(chan V)
+	}
+	notifee := &Observable[V]{sync.RWMutex{}, notifier, []chan V{}, false}
+
+	// Start listening to the notifier and emit values to subscribers
+	go notifee.listen(notifier)
+
+	return notifee, notifier
+}
+
+// Subscribe gets a subscription to the observable.
+func (obs *Observable[V]) Subscribe(ctx context.Context) observable.Subscription[V] {
+	obs.mu.Lock()
+	defer obs.mu.Unlock()
+
+	// Create a channel for the subscriber and append it to the subscribers list
+	ch := make(chan V, 1)
+	obs.subscribers = append(obs.subscribers, ch)
+
+	// Removal function used when unsubscribing from the observable
+	removeFromObservable := func() {
+		obs.mu.Lock()
+		defer obs.mu.Unlock()
+
+		for i, s := range obs.subscribers {
+			if ch == s {
+				obs.subscribers = append(obs.subscribers[:i], obs.subscribers[i+1:]...)
+				break
+			}
+		}
+	}
+
+	// Subscription gets its closed state from the observable
+	subscription := &Subscription[V]{ch, obs.closed, removeFromObservable}
+
+	go unsubscribeOnDone[V](ctx, subscription)
+
+	return subscription
+}
+
+// listen to the notifier and notify subscribers when values are received. This
+// function is blocking and should be run in a goroutine.
+func (obs *Observable[V]) listen(notifier <-chan V) {
```

Comment:
I gave it some thought and gave renaming `Subscription` to `Observer` a try. I also had [this conversation w/ chatGPT](https://chat.openai.com/share/a76baee4-537e-4455-8404-2ad0e58163db) and am now considering doing the following as well:

1. rename `notifier` to `producer`
2. rename the `notifier` package to `channel`
3. rename `notifiableObservable` to `channelObservable`
4. rename any references called `notifee` to `observable` (or something similar)

Reply:
Are you still contemplating these? I'd personally be in support of not having the `notifier`/`notifiee` naming convention as it's too generic IMO but don't want to block things too much.

Also, read through your thread, for future reference, consider just giving it the code rather than explaining the types & names you have:  https://chat.openai.com/share/d901a206-f152-4230-9c85-59ec08cde8fb